OpenRA.Mods.Common/Activities/Air/Fly.cs

```diff
@@ -51,34 +50,95 @@ public Fly(Actor self, Target t, WPos? initialTargetPosition = null, Color? targ
+			if (dat != desiredAltitude || move.Z != 0)
+			{
+				var maxDelta = move.HorizontalLength * aircraft.Info.MaximumPitch.Tan() / 1024;
+				var moveZ = move.Z != 0 ? move.Z : (desiredAltitude.Length - dat.Length);
+				var deltaZ = moveZ.Clamp(-maxDelta, maxDelta);
+				move = new WVec(move.X, move.Y, deltaZ);
+			}
+
+			aircraft.SetPosition(self, aircraft.CenterPosition + move);
+		}
+
+		public static void FlyTick(Actor self, Aircraft aircraft, int desiredFacing, WDist desiredAltitude, int turnSpeedOverride = -1)
+		{
+			FlyTick(self, aircraft, desiredFacing, desiredAltitude, WVec.Zero, turnSpeedOverride);
+		}
+
+		// Should only be used for vertical-only movement, usually VTOL take-off or land. Terrain-induced altitude changes should always be handled by FlyTick.
+		public static bool VerticalTakeOffOrLandTick(Actor self, Aircraft aircraft, int desiredFacing, WDist desiredAltitude, int turnSpeedOverride = -1)
+		{
+			var dat = self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition);
+			var move = WVec.Zero;
 
 			var turnSpeed = turnSpeedOverride > -1 ? turnSpeedOverride : aircraft.TurnSpeed;
 			aircraft.Facing = Util.TickFacing(aircraft.Facing, desiredFacing, turnSpeed);
 
-			if (altitude != desiredAltitude.Length)
+			if (dat != desiredAltitude)
 			{
-				var delta = move.HorizontalLength * aircraft.Info.MaximumPitch.Tan() / 1024;
-				var dz = (desiredAltitude.Length - altitude).Clamp(-delta, delta);
-				move += new WVec(0, 0, dz);
+				var maxDelta = aircraft.Info.AltitudeVelocity.Length;
+				var deltaZ = (desiredAltitude.Length - dat.Length).Clamp(-maxDelta, maxDelta);
+				move += new WVec(0, 0, deltaZ);
 			}
+			else
+				return false;
 
 			aircraft.SetPosition(self, aircraft.CenterPosition + move);
+			return true;
 		}
 
 		public override Activity Tick(Actor self)
 		{
+			if (ChildActivity != null)
+			{
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+				if (ChildActivity != null)
+					return this;
+			}
+
 			// Refuse to take off if it would land immediately again.
 			if (aircraft.ForceLanding)
 				Cancel(self);
 
+			var dat = self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition);
 			if (IsCanceling)
+			{
+				// We must return the actor to a sensible height before continuing.
+				// If the aircraft is on the ground we queue TakeOff to manage the influence reservation and takeoff sounds etc.
+				// TODO: It would be better to not take off at all, but we lack the plumbing to detect current airborne/landed state.
+				// If the aircraft lands when idle and is idle, we let the default idle handler manage this.
+				// TODO: Remove this after fixing all activities to work properly with arbitrary starting altitudes.
+				var skipHeightAdjustment = aircraft.Info.LandWhenIdle && self.CurrentActivity.IsCanceling && self.CurrentActivity.NextActivity == null;
+				if (!skipHeightAdjustment && dat != aircraft.Info.CruiseAltitude)
```

Comment:
Add `aircraft.Info.CanHover &&` here to fix Banshees/Orca Bombers doing a VTOL step when stopped/issued a new order just after a cliff height change.